### PR TITLE
Refactor some loops code and hook tests

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Assortment of minor changes
- Joined hook order tests
- Added hook order test for fit when no validation is run
- Tiny bit of typing
- Simplified the "should run validation" logic
- Simplified the update lr logic
- A test refactor

Reducing the diff for #7357

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -897,14 +894,20 @@ def should_accumulate(self):
...
+        is_val_check_epoch = (self.trainer.current_epoch + 1) % self.trainer.check_val_every_n_epoch == 0
+        if not is_val_check_epoch:
             return False
 
+        is_infinite_dataset = self.trainer.val_check_batch == float('inf')
+        if is_last_batch and is_infinite_dataset:
+            return True
+
+        if self.trainer.should_stop:
+            return True
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:909]: This is open to discussion, as it previously ran under `if on_epoch`

File: tests/models/test_hooks.py [hunk 2]
```diff
@@ -419,43 +460,25 @@ def teardown(self, stage=None):
...
-        'on_after_backward',
-        'on_train_batch_end',
-        'on_train_batch_start',
-        'on_before_batch_transfer',
-        'transfer_batch_to_device',
-        'on_after_batch_transfer',
-        'on_before_zero_grad',
-        'on_after_backward',
-        'on_train_batch_end',
+        *(model.train_batch * train_batches),
```
Comment (awaelchli) [inline tests/models/test_hooks.py:472]: the deleted lines do not match what is defined in model.train_batch
can you give a hint what that means?

Comment (carmocca) [reply tests/models/test_hooks.py:472]: Opened https://github.com/PyTorchLightning/pytorch-lightning/pull/7689 to just move the class so it's easier to review

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:909]: Let's add it to the changelog as it is not strictly a refactor?

File: pytorch_lightning/trainer/connectors/optimizer_connector.py [hunk 3]
```diff
@@ -11,30 +11,30 @@
...
 from pytorch_lightning.utilities import rank_zero_warn
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 
 
 class OptimizerConnector:
 
-    def __init__(self, trainer):
-        self.trainer = trainer
+    def __init__(self, trainer: 'pl.Trainer') -> None:
+        self.trainer = proxy(trainer)
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/optimizer_connector.py:25]: just curious what this proxy does...

File: tests/models/test_hooks.py [hunk 4]
```diff
@@ -234,11 +234,55 @@ class HookedModel(BoringModel):
...
+        ]
+        self.val_batch = [
+            'on_validation_batch_start',
+            'on_before_batch_transfer',
+            'transfer_batch_to_device',
+            'on_after_batch_transfer',
+            'on_validation_batch_end',
+        ]
+
+    def training_step(self, *args, **kwargs):
```
Comment (Borda) [inline tests/models/test_hooks.py:258]: wahy not use Mock? @awaelchli

File: tests/models/test_hooks.py [hunk 5]
```diff
@@ -463,14 +490,45 @@ def test_trainer_model_hook_system_fit(tmpdir):
...
+    trainer.fit(model)
+    expected = [
+        'setup_fit',
+        'on_fit_start',
+        'on_pretrain_routine_start',
+        'on_pretrain_routine_end',
+        'on_train_start',
+        'on_epoch_start',
+        'on_train_epoch_start',
+        *(model.train_batch * train_batches),
```
Comment (Borda) [inline tests/models/test_hooks.py:514]: :+1:

File: pytorch_lightning/trainer/connectors/optimizer_connector.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/optimizer_connector.py:25]: it keeps a weak reference to trainer without increasing the actual ref count, meaning garbage collector can delete trainer even though the proxy is still holding the trainer. Something like that.

File: tests/models/test_hooks.py (hunk 4 above)
Comment (carmocca) [reply tests/models/test_hooks.py:258]: It's hard to use as there are several checks throughout our codebase for `isinstance(LightningModule)`

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:909]: Actually, going to keep the `on_epoch` guard and address this separately.

```suggestion
        if on_epoch and self.trainer.should_stop:
            return True
```

**Review by awaelchli — APPROVED** [review]

File: tests/models/test_hooks.py (hunk 4 above)
Comment (Borda) [reply tests/models/test_hooks.py:258]: ok, so lest check if we have a ticket for mocking this kind of tests and if note create one...

**Review by Borda — APPROVED** [review]

File: tests/models/test_hooks.py (hunk 4 above)
Comment (carmocca) [reply tests/models/test_hooks.py:258]: I don't think it's possible, maybe if we could mock the inherited classes defining all these hooks...

**Pull request merged**